File: tool/upload_bibles.dart

import 'dart:convert';
import 'dart:developer' as developer;
import 'dart:io';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/widgets.dart';
import 'package:kyros/firebase_options.dart';

// IMPORTANT: This script MUST be run using the following command:
// flutter run -t tool/upload_bibles.dart

Future<void> main() async {
  // This ensures all Flutter and Firebase services are correctly initialized.
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );

  developer.log("Firebase App Initialized successfully. Starting upload...");

  final firestore = FirebaseFirestore.instance;
  final versionsDirectory = Directory('assets/translations/EN-English');

  if (!await versionsDirectory.exists()) {
    developer.log('Error: Directory not found: ${versionsDirectory.path}');
    return;
  }

  final files = versionsDirectory.listSync().where((item) => item.path.endsWith('.json'));

  for (final file in files) {
    try {
      final versionId = file.path.split('/').last.split('.').first.toUpperCase();
      final content = await File(file.path).readAsString();
      final jsonData = json.decode(content);

      // 1. Upload Version Metadata
      final versionDocRef = firestore.collection('versions').doc(versionId);
      await versionDocRef.set({
        'name': _getVersionName(versionId),
        'language': 'en',
        'fileName': file.path.split('/').last,
      });

      developer.log('Uploaded metadata for $versionId');

      // 2. Upload Bible Content (Book by Book)
      final books = jsonData['books'] as List;
      final bibleCollectionRef = firestore.collection('bibles').doc(versionId).collection('books');

      for (final bookData in books) {
        final bookName = bookData['name'];
        final bookChapters = bookData['chapters'] as List;

        final chaptersMap = <String, dynamic>{};
        for (int i = 0; i < bookChapters.length; i++) {
          final chapterNum = (i + 1).toString();
          chaptersMap[chapterNum] = bookChapters[i];
        }

        await bibleCollectionRef.doc(bookName).set({'chapters': chaptersMap});
        developer.log('  - Uploaded book: $bookName', name: versionId);
      }

      developer.log('Successfully uploaded all books for version $versionId.');

    } catch (e, s) {
      developer.log('Error processing file ${file.path}', error: e, stackTrace: s);
    }
  }

  developer.log('Bible upload process completed.');
  // Exit the app after the script is done.
  exit(0);
}

String _getVersionName(String versionId) {
  switch (versionId.toLowerCase()) {
    case 'asv': return 'American Standard Version';
    case 'asvs': return '''American Standard Version (Strong's)''';
    case 'bishops': return '''Bishops' Bible''';
    case 'coverdale': return 'Coverdale Bible';
    case 'geneva': return 'Geneva Bible';
    case 'kjv': return 'King James Version';
    case 'kjv_strongs': return '''King James Version (Strong's)''';
    case 'net': return 'New English Translation';
    case 'tyndale': return 'Tyndale Bible';
    case 'web': return 'World English Bible';
    default: return versionId;
  }
}

File: tool/upload_bibles.js
const admin = require('firebase-admin');
const fs = require('fs').promises;
const path = require('path');

// Initialize Firebase Admin SDK
const serviceAccount = require('./serviceAccountKey.json');
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});
const db = admin.firestore();

async function uploadBibles() {
    try {
        const versionsDirectory = path.join(__dirname, '../assets/translations/EN-English');
        console.log(`Reading from directory: ${versionsDirectory}`);

        const files = await fs.readdir(versionsDirectory);
        const jsonFiles = files.filter(file => path.extname(file) === '.json');

        for (const file of jsonFiles) {
            const versionId = path.basename(file, '.json').toUpperCase();
            const filePath = path.join(versionsDirectory, file);

            try {
                const content = await fs.readFile(filePath, 'utf8');
                const jsonData = JSON.parse(content);

                // 1. Upload Version Metadata
                const versionDocRef = db.collection('versions').doc(versionId);
                await versionDocRef.set({
                    name: getVersionName(versionId),
                    language: 'en',
                    fileName: file,
                });
                console.log(`Uploaded metadata for ${versionId}`);

                // 2. Upload Bible Content (Book by Book)
                const bibleCollectionRef = db.collection('bibles').doc(versionId).collection('books');

                const books = jsonData.books;

                for (const bookData of books) {
                    const bookName = bookData.name;
                    const bookChapters = bookData.chapters;

                    const chaptersMap = {};
                    for (let i = 0; i < bookChapters.length; i++) {
                        const chapterNum = (i + 1).toString();
                        chaptersMap[chapterNum] = bookChapters[i];
                    }

                    await bibleCollectionRef.doc(bookName).set({ chapters: chaptersMap });
                    console.log(`  - Uploaded book: ${bookName}`);
                }

                console.log(`Successfully uploaded all books for version ${versionId}.`);
            } catch (e) {
                console.error(`Error processing file ${file}`, e);
            }
        }

        console.log('Bible upload process completed.');
    } catch (error) {
        console.error('Error uploading bibles:', error);
    }
}

function getVersionName(versionId) {
    // Simple mapping, can be extended
    const names = {
        'AKJV_BIBLE': 'American King James Version',
        'AMP_BIBLE': 'Amplified Bible',
        'ASV_BIBLE': 'American Standard Version',
        'BRG_BIBLE': 'BRG Bible',
        'CSB_BIBLE': 'Christian Standard Bible',
        'EHV_BIBLE': 'Evangelical Heritage Version',
        'ESVUK_BIBLE': 'English Standard Version Anglicised',
        'ESV_BIBLE': 'English Standard Version',
        'GNV_BIBLE': 'Geneva Bible',
        'GW_BIBLE': 'GOD\'S WORD Translation',
        'ISV_BIBLE': 'International Standard Version',
        'JUB_BIBLE': 'Jubilee Bible 2000',
        'KJ21_BIBLE': '21st Century King James Version',
        'KJV_BIBLE': 'King James Version',
        'LEB_BIBLE': 'Lexham English Bible',
        'MEV_BIBLE': 'Modern English Version',
        'NASB1995_BIBLE': 'New American Standard Bible 1995',
        'NASB_BIBLE': 'New American Standard Bible',
        'NET_BIBLE': 'New English Translation',
        'NIVUK_BIBLE': 'New International Version - UK',
        'NIV_BIBLE': 'New International Version',
        'NKJV_BIBLE': 'New King James Version',
        'NLT_BIBLE': 'New Living Translation',
        'NLV_BIBLE': 'New Life Version',
        'NOG_BIBLE': 'Names of God Bible',
        'NRSVUE_BIBLE': 'New Revised Standard Version Updated Edition',
        'NRSV_BIBLE': 'New Revised Standard Version',
        'WEB_BIBLE': 'World English Bible',
        'YLT_BIBLE': 'Young\'s Literal Translation'
    };
    return names[versionId] || versionId.replace('_BIBLE', '');
}

uploadBibles();